src/main/scala/mesosphere/marathon/core/matcher/manager/impl/OfferMatcherManagerActor.scala

```diff
@@ -0,0 +1,222 @@
+import org.slf4j.LoggerFactory
+import rx.lang.scala.Observer
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Queue
+import scala.util.Random
+import scala.util.control.NonFatal
+
+/**
+  * This actor offers one interface to a dynamic collection of matchers
+  * and includes logic for limiting the amount of launches.
+  */
+private[manager] object OfferMatcherManagerActor {
+  def props(
+    random: Random, clock: Clock,
+    offerMatcherConfig: OfferMatcherConfig, offersWanted: Observer[Boolean]): Props = {
+    Props(new OfferMatcherManagerActor(random, clock, offerMatcherConfig, offersWanted))
+  }
+
+  private val log = LoggerFactory.getLogger(getClass)
+  private case class OfferData(
+      offer: Offer,
+      deadline: Timestamp,
+      sender: ActorRef,
+      matcherQueue: Queue[OfferMatcher],
+      tasks: Seq[TaskWithSource],
+      matchPasses: Int = 0) {
+
+    def addMatcher(matcher: OfferMatcher): OfferData = copy(matcherQueue = matcherQueue.enqueue(matcher))
+    def nextMatcherOpt: Option[(OfferMatcher, OfferData)] = {
+      matcherQueue.dequeueOption map {
+        case (nextMatcher, newQueue) => nextMatcher -> copy(matcherQueue = newQueue)
+      }
+    }
+
+    def addTasks(added: Seq[TaskWithSource]): OfferData = {
+      val offerResources: Seq[Resource] = offer.getResourcesList.asScala
+      val taskResources: Seq[Resource] = added.map(_.taskInfo).flatMap(_.getResourcesList.asScala)
+      val leftOverResources = ResourceUtil.consumeResources(offerResources, taskResources)
+      val leftOverOffer = offer.toBuilder.clearResources().addAllResources(leftOverResources.asJava).build()
+      copy(
+        offer = leftOverOffer,
+        tasks = added ++ tasks
+      )
+    }
+  }
+}
+
+private class OfferMatcherManagerActor private (
+  random: Random, clock: Clock, conf: OfferMatcherConfig, offersWantedObserver: Observer[Boolean])
+    extends Actor with ActorLogging {
+  private[this] var launchTokens: Int = 0
+
+  private[this] var matchers: Set[OfferMatcher] = Set.empty
+
+  private[this] var offerQueues: Map[OfferID, OfferMatcherManagerActor.OfferData] = Map.empty
+
+  override def receive: Receive = LoggingReceive {
+    Seq[Receive](
+      receiveSetLaunchTokens,
+      receiveChangingMatchers,
+      receiveProcessOffer,
+      receiveMatchedTasks
+    ).reduceLeft(_.orElse[Any, Unit](_))
+  }
+
+  private[this] def receiveSetLaunchTokens: Receive = {
+    case ActorOfferMatcherManager.SetTaskLaunchTokens(tokens) =>
+      launchTokens = tokens
+      if (tokens > 0 && launchTokens == tokens)
+        updateOffersWanted()
+    case ActorOfferMatcherManager.AddTaskLaunchTokens(tokens) =>
+      launchTokens += tokens
+      if (tokens > 0 && launchTokens == tokens)
+        updateOffersWanted()
+  }
+
+  private[this] def receiveChangingMatchers: Receive = {
+    case ActorOfferMatcherManager.AddOrUpdateMatcher(matcher) =>
+      if (!matchers(matcher)) {
```

Comment:
Make the ActorOfferMatcher a case class, so you have proper hashCode/equals or is relying on reference enough here?